crossbeam-epoch/src/lib.rs

```diff
@@ -66,6 +66,87 @@ extern crate core;
 
 extern crate maybe_uninit;
 
+#[cfg(loom)]
+extern crate loom;
+
+#[cfg(loom)]
+#[allow(unused_imports, dead_code)]
+pub(crate) mod concurrency {
+    pub(crate) mod cell {
+        pub(crate) use loom::cell::UnsafeCell;
+    }
+    pub(crate) mod sync {
+        pub(crate) mod atomic {
+            pub(crate) use loom::sync::atomic::{AtomicUsize, Ordering};
+            pub(crate) fn fence(ord: Ordering) {
+                if let Ordering::Acquire = ord {
+                } else {
+                    // FIXME: loom only supports acquire fences at the moment.
+                    // https://github.com/tokio-rs/loom/issues/117
+                    // let's at least not panic...
+                    // this may generate some false positives (`SeqCst` is stronger than `Acquire`
+                    // for example), and some false negatives (`Relaxed` is weaker than `Acquire`),
+                    // but it's the best we can do for the time being.
+                }
+                loom::sync::atomic::fence(Ordering::Acquire)
+            }
+
+            // FIXME: loom does not support compiler_fence at the moment.
+            // https://github.com/tokio-rs/loom/issues/117
+            // we use fence as a stand-in for compiler_fence for the time being.
+            // this may miss some races since fence is stronger than compiler_fence,
+            // but it's the best we can do for the time being.
+            pub(crate) use self::fence as compiler_fence;
+        }
+        pub(crate) use loom::sync::Arc;
+    }
+    pub(crate) use loom::lazy_static;
+    pub(crate) use loom::thread_local;
+}
+#[cfg(not(loom))]
+#[allow(unused_imports, dead_code)]
+pub(crate) mod concurrency {
+    #[cfg(any(feature = "alloc", feature = "std"))]
+    pub(crate) mod cell {
+        #[derive(Debug)]
+        #[repr(transparent)]
+        pub(crate) struct UnsafeCell<T>(::core::cell::UnsafeCell<T>);
```

Comment:
Why this is a new type, not just reexport?

Reply:
I'm not sure I follow — I need to add new methods to it (`with` and `with_mut`), which requires a newtype.

Reply:
Oops, sorry! I somehow missed that UnsafeCell does not have those methods.

Reply:
Yeah, loom's `UnsafeCell` has a slightly different API basically because it wants to track how long you use the mutable reference for. And so we need to shim that in the non-loom case.